### PR TITLE
Implement `%%` and `%%=`

### LINKED ISSUE #853: ` %%` and `%%=` operators not implemented
> This is a newer version of the % operator that supports all numbers, not just integers. It is equivalent to B * fract(A / B). The % operator does the same thing, but truncates A and B to integers first.

https://www.byond.com/docs/ref/#/operator/%25%25
https://www.byond.com/docs/ref/#/operator/%25%25=

### PR DESCRIPTION
**Pull request by ike709**
Fixes #853 

I've done some very basic testing and it seems to work.

### CODE REVIEW

**Comment by ZeWaka** [comment]
Requires #883

**Comment by ike709** [comment]
No it doesn't, this won't impact 514 codebases at all.

File: Content.Tests/DMProject/Tests/Operators/modulo_modulo.dm [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+/proc/RunTest()
+	ASSERT(5.89254 %% 2.3462 == 1.2001398)
+	var/A = 5.89254
+	A %%= 2.3462
+	ASSERT(A == 1.2001398)
```
Comment (wixoaGit) [inline Content.Tests/DMProject/Tests/Operators/modulo_modulo.dm:6]: Both these assertions fail on BYOND

File: DMCompiler/DM/Expressions/Constant.cs [hunk 2]
```diff
@@ -173,6 +177,17 @@ class Number : Constant {
             return new Number(Location, Value % rhsNum.Value);
         }
 
+        public override Constant ModuloModulo(Constant rhs) {
+            if (rhs is not Number rhsNum) {
+                return base.Add(rhs);
```
Comment (wixoaGit) [inline DMCompiler/DM/Expressions/Constant.cs:182]: This probably means to call `base.ModuloMudulo()`

..So does every other math operation override that seems to have copy-pasted `base.Add()`
```suggestion
                return base.ModuloModulo(rhs);
```

File: DMCompiler/DM/Visitors/DMVisitorExpression.cs [hunk 3]
```diff
@@ -321,6 +327,12 @@ public void VisitIdentifier(DMASTIdentifier identifier)
             Result = new Expressions.ModulusAssign(modulusAssign.Location, lhs, rhs);
         }
 
+        public void VisitModulusModulusAssign(DMASTModulusModulusAssign modulusModulusAssign) {
+            var lhs = DMExpression.Create(_dmObject, _proc, modulusModulusAssign.A);
+            var rhs = DMExpression.Create(_dmObject, _proc, modulusModulusAssign.B);
```
Comment (wixoaGit) [inline DMCompiler/DM/Visitors/DMVisitorExpression.cs:332]: You can do `a %%= new`
```suggestion
            var lhs = DMExpression.Create(_dmObject, _proc, modulusModulusAssign.A, _inferredPath);
            var rhs = DMExpression.Create(_dmObject, _proc, modulusModulusAssign.B, lhs.Path);
```

File: OpenDreamRuntime/Procs/DMOpcodeHandlers.cs [hunk 4]
```diff
@@ -951,6 +951,19 @@ private static void HandleSuffixPronoun(ref StringBuilder formattedString, ReadO
...
 
+        public static ProcStatus? ModulusModulus(DMProcState state) {
+            DreamValue second = state.Pop();
+            DreamValue first = state.Pop();
+
+            if (first.Type == DreamValue.DreamValueType.Float && second.Type == DreamValue.DreamValueType.Float) {
+                state.Push(ModulusModulusValues(first, second));
+            } else {
+                throw new Exception("Invalid modulusmodulus operation on " + first + " and " + second);
+            }
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/DMOpcodeHandlers.cs:962]: `ModulusModulusValues()` already does the type check and exception
```suggestion
            state.Push(ModulusModulusValues(first, second));
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: Content.Tests/DMProject/Tests/Operators/modulo_modulo.dm [hunk 5]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (ike709) [inline Content.Tests/DMProject/Tests/Operators/modulo_modulo.dm:1]:
```suggestion
// BREAKING CHANGE: The floating point precision is slightly different between OD and BYOND, giving slightly different values
```

File: OpenDreamRuntime/Procs/DMOpcodeHandlers.cs [hunk 6]
```diff
@@ -2002,6 +2022,16 @@ private static void HandleSuffixPronoun(ref StringBuilder formattedString, ReadO
                 throw new Exception("Invalid modulus operation on " + first + " and " + second);
             }
         }
+
+        private static DreamValue ModulusModulusValues(DreamValue first, DreamValue second) {
+            if (first.TryGetValueAsFloat(out var firstFloat) && second.TryGetValueAsFloat(out var secondFloat)) {
+                // BYOND docs say that A %% B is equivalent to B * fract(A/B)
```
Comment (ike709) [inline OpenDreamRuntime/Procs/DMOpcodeHandlers.cs:2028]:
```suggestion
                // BYOND docs say that A %% B is equivalent to B * fract(A/B)
                // BREAKING CHANGE: The floating point precision is slightly different between OD and BYOND, giving slightly different values
```

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**